mlir/lib/Dialect/Rock/Transforms/SugarToLoops.cpp

```diff
@@ -1085,6 +1097,64 @@ static Value zeroDMemrefAsOneD(PatternRewriter &b, Value memref) {
                                                memref, reassociationRef);
 }
 
+std::tuple<SmallVector<Value>, Type> getCoordsAndType(PatternRewriter &b,
+                                                      GlobalLoadOp op) {
+  MemRefType srcType = op.getSource().getType();
+  Location loc = op.getLoc();
+  SmallVector<Value> coords(op.getSourceCoord());
+  Type originalLoadedType = op.getResult().getType();
+  int64_t originalLoadVecLen = 1;
+  if (VectorType originalLoadVecType =
+          dyn_cast<VectorType>(originalLoadedType)) {
+    originalLoadVecLen = originalLoadVecType.getNumElements();
+  }
+  if (srcType.getElementType().getIntOrFloatBitWidth() >= 8 ||
+      originalLoadVecLen != 1) {
+    return {coords, originalLoadedType};
+  } else if (srcType.getElementType().getIntOrFloatBitWidth() == 4) {
+    ArrayRef<int64_t> shape = srcType.getShape();
+    Value flatAddress = flattenCoords(b, loc, coords, shape);
+    Type coordType = flatAddress.getType();
+    Value one = getConstIntOrIndexValue(b, loc, 1, coordType);
+    flatAddress = b.createOrFold<arith::ShRUIOp>(loc, flatAddress, one);
+    flatAddress = b.createOrFold<arith::ShLIOp>(loc, flatAddress, one);
+    SmallVector<Value> newCoords;
+    unflattenCoords(b, loc, flatAddress, shape, newCoords);
+    Type loadedType = VectorType::get({2}, srcType.getElementType());
+    return {newCoords, loadedType};
+  } else {
+    llvm_unreachable("less than 4 bit element types are not implemented");
+  }
+}
+
+Value selectDataIf4b(PatternRewriter &b, GlobalLoadOp op, Value loadedVec) {
+  MemRefType srcType = op.getSource().getType();
+  Type originalLoadedType = op.getResult().getType();
+  if (srcType.getElementType().getIntOrFloatBitWidth() >= 8) {
+    return loadedVec;
+  }
+  int64_t originalLoadVecLen = 1;
+  if (VectorType originalLoadVecType =
+          dyn_cast_if_present<VectorType>(originalLoadedType)) {
+    originalLoadVecLen = originalLoadVecType.getNumElements();
+  }
+  if (originalLoadVecLen != 1) {
+    return loadedVec;
+  } else if (srcType.getElementType().getIntOrFloatBitWidth() == 4) {
+    assert(isa<VectorType>(loadedVec.getType()));
+    Location loc = op.getLoc();
+    SmallVector<Value, 5> coords(op.getSourceCoord());
+    ArrayRef<int64_t> shape = srcType.getShape();
+    Value flatAddress = flattenCoords(b, loc, coords, shape);
+    Type coordType = flatAddress.getType();
+    Value one = getConstIntOrIndexValue(b, loc, 1, coordType);
+    Value lsb = b.createOrFold<arith::AndIOp>(loc, flatAddress, one);
```

Comment:
not really me thinks...
So the instruction that d remain after folding would be two `and` ops.

1) Above: i ve changed the above (I ll push them next) to use lsbMask which is not the same as exracting the lsb. (e.g. : ... 1110)
2) Below : here it would be just extracting (e.g 0001)

So at least there need to be two instructions.

Reply:
Nice, thanks! I wasn't aware of how folding works, completely ignored the fold part of createOrFold :D